segmenter/video_segmenter.go

```diff
@@ -0,0 +1,223 @@
+
+	//Invoke the FFMpeg command
+	// fmt.Println("ffmpeg", "-i", fmt.Sprintf("rtmp://localhost:%v/stream/%v", "1935", "test"), "-vcodec", "copy", "-acodec", "copy", "-bsf:v", "h264_mp4toannexb", "-f", "segment", "-muxdelay", "0", "-segment_list", "./tmp/stream.m3u8", "./tmp/stream_%d.ts")
+	plfn := fmt.Sprintf("%s/%s.m3u8", s.WorkDir, s.StrmID)
+	tsfn := s.WorkDir + "/" + s.StrmID + "_%d.ts"
+
+	//This command needs to be manually killed, because ffmpeg doesn't seem to quit after getting a rtmp EOF
+	cmd := exec.Command("ffmpeg", "-i", s.LocalRtmpUrl, "-vcodec", "copy", "-acodec", "copy", "-bsf:v", "h264_mp4toannexb", "-f", "segment", "-muxdelay", "0", "-segment_list", plfn, tsfn)
+	err = cmd.Start()
+	if err != nil {
+		glog.Errorf("Cannot start ffmpeg command.")
+		return err
+	}
+
+	ec := make(chan error, 1)
+	go func() { ec <- cmd.Wait() }()
+
+	select {
+	case ffmpege := <-ec:
+		glog.Errorf("Error from ffmpeg: %v", ffmpege)
+		return ffmpege
+	case <-ctx.Done():
+		//Can't close RTMP server, joy4 doesn't support it.
+		//server.Stop()
+		cmd.Process.Kill()
+		return ctx.Err()
+	}
+}
+
+//PollSegment monitors the filesystem and returns a new segment as it becomes available
+func (s *FFMpegVideoSegmenter) PollSegment(ctx context.Context) (*VideoSegment, error) {
+	var length time.Duration
+	tsfn := s.WorkDir + "/" + s.StrmID + "_" + strconv.Itoa(s.curSegment) + ".ts"
+	seg, err := pollSegment(ctx, tsfn, time.Millisecond*100, s.SegLen)
+	if err != nil {
+		return nil, err
+	}
+
+	name := s.StrmID + "_" + strconv.Itoa(s.curSegment) + ".ts"
+	if s.curPlaylist != nil && s.curPlaylist.Segments[s.curSegment] != nil {
+		//This is ridiculous - but it's how we can round floats in Go
+		sec, _ := strconv.Atoi(fmt.Sprintf("%.0f", s.curPlaylist.Segments[s.curSegment].Duration))
+		length = time.Duration(sec) * 1000 * time.Millisecond
+	}
+
+	s.curSegment = s.curSegment + 1
+	glog.Infof("Segment: %v, len:%v", name, len(seg))
+	return &VideoSegment{Codec: av.H264, Format: HLS, Length: length, Data: seg, Name: name}, err
+}
+
+//PollPlaylist monitors the filesystem and returns a new playlist as it becomes available
+func (s *FFMpegVideoSegmenter) PollPlaylist(ctx context.Context) (*VideoPlaylist, error) {
+	plfn := fmt.Sprintf("%s/%s.m3u8", s.WorkDir, s.StrmID)
+	var lastPl []byte
+	if s.curPlaylist == nil {
+		lastPl = nil
+	} else {
+		lastPl = s.curPlaylist.Encode().Bytes()
+	}
+
+	pl, err := pollPlaylist(ctx, plfn, time.Millisecond*100, lastPl)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := m3u8.NewMediaPlaylist(50000, 50000)
+	err = p.DecodeFrom(bytes.NewReader(pl), true)
+	if err != nil {
+		return nil, err
+	}
+
+	s.curPlaylist = p
+	return &VideoPlaylist{Format: HLS, Data: p}, err
+}
+
+func pollPlaylist(ctx context.Context, fn string, sleepTime time.Duration, lastFile []byte) (f []byte, err error) {
+	for {
+		if _, err := os.Stat(fn); err == nil {
+			if err != nil {
+				return nil, err
+			}
+
+			content, err := ioutil.ReadFile(fn)
```

Comment:
Not right now... I agree it can be a problem - was thinking about adding a "clear history" button in the UI or something...